src/cluster.c

```diff
@@ -5793,63 +5953,97 @@ NULL
+                                         CLUSTER_TODO_FSYNC_CONFIG);
+                }
+
+                /* If this node or this node's primary was importing this slot,
+                * assigning the slot to itself also clears the importing status. */
+                if ((n == myself || n == myself->slaveof) &&
+                    server.cluster->importing_slots_from[slot])
+                {
+                    server.cluster->importing_slots_from[slot] = NULL;
+
+                    /* Only primary broadcasts the updates */
+                    if (n == myself) {
+                        /* This slot was manually migrated, set this node configEpoch
+                        * to a new epoch so that the new version can be propagated
+                        * by the cluster.
+                        *
+                        * Note that if this ever results in a collision with another
+                        * node getting the same configEpoch, for example because a
+                        * failover happens at the same time we close the slot, the
+                        * configEpoch collision resolution will fix it assigning
+                        * a different epoch to each node. */
+                        if (clusterBumpConfigEpochWithoutConsensus() == C_OK) {
+                            serverLog(LL_WARNING,
+                                      "ConfigEpoch updated after importing slot %d",
+                                      slot);
+                        }
+                        /* After importing this slot, let the other nodes know as
+                        * soon as possible. */
+                        clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+                    }
                 }
-                server.cluster->importing_slots_from[slot] = NULL;
-                /* After importing this slot, let the other nodes know as
-                 * soon as possible. */
-                clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+
+                /* Don't replicate again if setslot has been replicated */
+                replSetSlot &= !replDone;
+            } else {
+                /* We are a primary and this is the first time we see this "setslot"
+                 * command. Force-replicate the setslot command to all of our replicas
+                 * first and only on success will we handle the command.
+                 * Note that
+                 * 1. All replicas are expected to ack the replication within 1000ms
+                 * 2. The repl offset target is set to the master's current repl offset + 1.
+                 *    There is no concern of partial replication because replicas always
+                 *    ack the repl offset at the command boundary. */
+                waitForReplication(c, mstime()+1000, myself->numslaves, server.master_repl_offset+1, BLOCKED_WAIT_RERUN);
+
+                /* Don't reply to the client yet */
+                reply = 0;
```

Comment:
I'm not entirely clear why this can't just be a `WAIT` command issued by the client for now. I feel like there is some interesting ideas here about having having deep support in Redis for commands that wait for replicas to process them before acknowledging and applying them locally, but I'm not sure we need that here.

Reply:
The functionality in question is similar to that of the WAIT command, but there is a crucial difference. The WAIT command is intended for use by real clients, so when it completes, the replication status must be returned to the client that issued the original WAIT command. In contrast, in this case, the server must consume the replication status and continue executing the pending command. Initially, I considered building the new WAIT* functionality on top of the existing WAIT command handling logic, but I ultimately concluded that it would be cleaner to introduce a new WAIT* contract instead.

Reply:
I want to make sure we've exhausted the solution space before starting to go down this path, I really don't want a pseudo synchronous mechanism hanging out in the engine given the complexity. My understanding is that we are doing this so that we handle cases where neither A or A' own the slot anymore but B died, but my understanding is we can reconcile those states without this replication mechanism by just having B take ownership after some threshold. That seems more inline with the general reconciliatory nature of Redis cluster. I also think it will be better handled in the cluster V2 world, since the TD will issue the updated state.

I believe this was discussed here: https://github.com/redis/redis/issues/6339#issuecomment-1088211873.

Reply:
so you need that because you want a command to be executed half way (propagating something, or initiating replication), then get blocked on a replication offset, and then when the repl_offset matches, the second half of the command is getting executed?
it seems a bit against the concept of "command" in redis (being atomic), maybe we should somehow split that into two commands and issue a WAIT in between (sending them in pipeline)?

i didn't dig into the changes in cluster.c, so correct me if i'm wrong.
if we somehow give that up, we can revert nearly / all changes to blocked.c, replication.c, server.h, and server.c.

Reply:
>so you need that because you want a command to be executed half way (propagating something, or initiating replication), then get blocked on a replication offset, and then when the repl_offset matches, the second half of the command is getting executed?

Not really. What I need is essentially "synchronous replication" such that the command is first executed on the replicas and only upon successful replication gets executed on the primary. 

>it seems a bit against the concept of "command" in redis (being atomic), maybe we should somehow split that into two commands and issue a WAIT in between (sending them in pipeline)?

The command is still executed as an atomic unit so atomicity is upheld. The new thing is the "synchronous replication", which has not been done before.

The command in question here is `cluster set <slot> <node>`. The idea is to maintain backward compatibility with the existing slot migration protocol which always sends the command to the primaries. An alternative would be to have the admin finalize the slot ownership on the replicas explicitly. This is unconventional IMO and places more burden on the admin side. 

I am not sure how to break the command apart into two parts. It is the execution order (primary after replicas) that this PR is trying to address.

>i didn't dig into the changes in cluster.c, so correct me if i'm wrong.

No worries. There are two SPOFs this PR is trying to solve. One happens during the slot migration and the other happens during the slot ownership finalization. The first one is mitigated by replicating the migration states to replicas either at the slot migration staring time and whenever a new replica comes up online. The second one is the one that requires the ownership to be finalized on the target replicas (if any) first, then on the target primary, and finally on the source shard, where the execution order between source primary and source replicas isn't material. This is the only order that would prevent a SPOF caused by a target primary crashing shortly after broadcasting its ownership to the cluster but target replicas getting stuck in the `IMPORTING` state.  This [link](https://github.com/redis/redis/issues/6339#issuecomment-1088211873) has the details for the second SPOF.